[GAWB-2142] Implement metrics for workflow and submission statuses

rtitle opened the pull request:
https://broadinstitute.atlassian.net/browse/GAWB-2142

Not quite ready yet as I'm still working on testing -- including end-to-end testing with statsd and stackdriver. I thought I'd open this PR early for review/feedback though.

The following metrics were added:
- Lifetime counts of submission status, broken out by workspace
  - Answers the question:
    -  _How many submissions were submitted/aborted/etc over the last N days [for a given workspace]?_
- Lifetime counts of workflow status, broken out by (workspace, submission)
  - Answers the question:
    - _How many workflows were submitted/queued/launched/aborted/etc over the last N days [for a given workspace and submission]?_
- Current counts of workflow statuses, broken out by (workspace, submission)
  - Answers the question:
    - _How many workflows are queued/running/aborting/etc **right now** [for a given workspace and submission]?_

The last one is more of a gauge of current system status, while the first 2 are more for looking at trends over time.

Note: reviewing tests with `?w=1` is recommended as there are a lot of whitespace changes with the addition of the pattern:
```
withStatsD {
  // test code
}
```

----------

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

rtitle committed:
Initial workflow/submission status instrumentation

rtitle committed:
Add counters to WorkflowSubmissionActor

rtitle committed:
Added workflow/submission status metrics to WorkspaceService

rtitle committed:
WorkflowSubmissionSpec tests

rtitle committed:
Fix merge issues after rebase

rtitle committed:
Fix test failures

rtitle committed:
Code cleanup

rtitle committed:
Fix compilation error

rtitle committed:
Added gauge metrics to SubmissionMonitorActor

rtitle committed:
Add clue for easier test debugging

rtitle committed:
Add some scaladoc to RawlsInstrumented

rtitle committed:
Use metricsPrefix in Boot.scala

rtitle committed:
Fix compilation error and warning:

rtitle committed:
SubmissionMonitorSpec test cases

rtitle committed:
Add cases to SubmissionApiServiceSpec

rtitle committed:
Minor tweaks

rtitle committed:
Fix compilation error, damn you Intellij

rtitle committed:
Add some DBIOAction shorthand

rtitle committed:
More idiomatic way of counting results of DBIOActions

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -0,0 +1,108 @@
+    * Adds a .countDBResult method to Counter which counts the result of a numeric DBIOAction.
+    */
+  protected implicit class CounterDBIOActionSupport(counter: Counter) {
+    def countDBResult[R <% Long, S <: NoStream, E <: Effect](action: DBIOAction[R, S, E])(implicit executionContext: ExecutionContext): DBIOAction[R, NoStream, E] =
```
IntelliJ gave me a deprecation warning here: https://issues.scala-lang.org/browse/SI-7629

rtitle replied:
View bounds are deprecated? News to me. There's probably another way to do this.

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:
```diff
@@ -110,7 +172,9 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging {
     def abortQueuedWorkflows(submissionId: UUID) = {
       datasource.inTransaction { dataAccess =>
-        dataAccess.workflowQuery.batchUpdateWorkflowsOfStatus(submissionId, WorkflowStatuses.Queued, WorkflowStatuses.Aborted)
+        workflowStatusCounter(WorkflowStatuses.Aborted).countDBResult {
```
What do you think about updating counters at the `WorkflowComponent` level?  That way every call to `batchUpdateWorkflowsOfStatus()` etc. would be instrumented without higher levels having to worry about it.

rtitle replied:
My thinking was that `WorkflowComponent` should only be concerned with updating the database, and that metric updates should be at the level above it (in the actual business logic). Plus, these metrics usually need a `Workspace` and a `Submission` which would need to be passed into `WorkflowComponent` solely for the purpose of updating metrics, which doesn't seem right.

Maybe instrumentation relating specifically to the database could live at the `*Component` layer (such as query timing, etc). But I felt these metrics are more closely tied to the actor logic.

rtitle replied:
Yeah, I see your point. I guess I was just hesitant to add side effects to what is now pure database-affecting code. But you're right it might make the code easier to work with. Let me take a stab at moving the counters to `WorkflowComponent` (there are still multiple methods that need updating) and see how it looks.

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:
```diff
@@ -123,20 +133,25 @@ trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths
+            val submissionId = workflowRecs.head.submissionId
+            val filteredWorkflowRecs = workflowRecs.filter(_.submissionId == submissionId)
+            for {
+              submissionRecOpt <- dataAccess.submissionQuery.findById(submissionId).result.map(_.headOption)
```
I like `headOption` for safety generally, but these are foreign key relations so Rawls should probably complain more loudly when this assumption fails

rtitle replied:
True - I saw `_.head` in other places. Maybe I'll change this to be consistent.

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -0,0 +1,108 @@
+    * Metric name fragments are combined via ExpandedMetricBuilder to generate an "expanded" metric name.
+    * By default this just calls toString on the object of type A, but this can be overridden.
+    */
+  @implicitNotFound(msg = "Cannot expand instances of type ${A}")
```
oh this is cool

rtitle replied:
Yeah it gives nicer compiler messages :)

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1428,7 +1435,11 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+          submissionStatusCounter(workspaceName, SubmissionStatuses.Submitted) += 1
+          workflowStatusCounter(workspaceName, UUID.fromString(submissionId), WorkflowStatuses.Queued) += workflows.size
+          sub
+        } map { _ =>
```
can we remove this line and also stop holding on to the `sub` ?

rtitle replied:
Yeah no need for the double-`map` I guess :)

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/StatsDTestUtils.scala:
```diff
@@ -0,0 +1,62 @@
+    try {
+      val result = testCode
+      eventually(timeout(10 seconds)) {
+        val order = mockitoInOrder(statsD)
```
I haven't used Mockito before.  Let me see if I have interpreted this correctly.

* record the calls to `statsD` methods in order
* verify that the calls were made in the order `connect()`, `send()` at least once, and then `close()`
* enable callers to access the arguments given to `send()` for further test verification

rtitle replied:
Yes that's exactly right. Furthermore it checks that `connect()` and `close()` were called exactly once, and it waits up to 10 seconds (polling every x milliseconds, whatever the default is) before giving up.

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/StatsDTestUtils.scala:
```diff
@@ -0,0 +1,62 @@
+    }
+  }
+
+  protected def expectedMetric(workspace: Workspace, submission: Submission, workflowStatus: WorkflowStatus, name: Option[String], expectedTimes: Int): (String, String) =
```
expectedWFStatusMetric

rtitle replied:
Yeah thanks

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala:
```diff
@@ -331,7 +331,7 @@ class RemoteServicesMockServer(port:Int) extends RawlsTestUtils {
         .withHeaders(jsonHeader)
         .withBody {
-          (1 to numWorkflows).map(_ => ExecutionServiceStatus(UUID.randomUUID().toString, "Submitted")).toList.toJson.toString
+          (1 to numWorkflows).map(n => ExecutionServiceStatus(UUID.randomUUID().toString, if (n == numWorkflows) "Failed" else "Submitted")).toList.toJson.toString
```
what is this change for?

rtitle replied:
Long story. :) 

This used to be hard coded to return 2 successes and 1 failure. In https://github.com/broadinstitute/rawls/pull/718 I changed it to be dynamic, but lost the failure. The tests still passed.

However when working on the instrumentation tests I found the test whose _intent_ it was to check for a failure -- the assertion was just weak (it was checking `status != Queued` or something). But while verifying instrumentation the test actually failed due to lack of a wf failure status. So I changed this logic to make the last workflow a failure in Cromwell.

jmthibault79 replied:
hoo boy.  :)

Could you add `// last one fails` or something to make the intention clear?

rtitle replied:
Yeah. 

BTW this is one of the reasons I like Mockito: it makes every test responsible for its own mocked data so you're not guessing which test uses which piece of a global test data set.

http://www.scalacheck.org/ is another nice tool for test data generation & property based testing -- I could give the team an overview sometime if interested.

geoffjentry replied:
Id love to see scalacheck start to get pulled in more

rtitle commented on core/src/test/scala/org/broadinstitute/dsde/rawls/StatsDTestUtils.scala:
```diff
@@ -0,0 +1,62 @@
+    try {
+      val result = testCode
+      eventually(timeout(10 seconds)) {
+        val order = mockitoInOrder(statsD)
```
Yes that's exactly right. Furthermore it checks that `connect()` and `close()` were called exactly once, and it waits up to 10 seconds (polling every x milliseconds, whatever the default is) before giving up.

rtitle replied:
Park street strikes again

rtitle committed:
Code review feedback

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:
```diff
@@ -77,17 +106,50 @@ class SubmissionMonitorActor(val workspaceName: WorkspaceName,
+    case CheckCurrentWorkflowStatusCounts =>
+      logger.debug(s"check current workflow status counts for submission $submissionId")
+      checkCurrentWorkflowStatusCounts pipeTo self
+    case SaveCurrentWorkflowStatusCounts(statusCounts) =>
```
Does this need to be a separate message at all? It's not doing much. Why not put it inside `checkCurrentWorkflowStatusCounts`?

rtitle replied:
Good question! Because `checkCurrentWorkflowStatusCounts` is running in a `Future` (using a thread from the `SlickDataSource` thread pool) and it's only safe to update actor state from the actor thread itself. So I pipe the database Future back to the actor and update the actor state upon receiving a `SaveCurrentWorkflowStatusCounts` message.

helgridly replied:
Oh, this is terrifying. Gonna look into this to wrap my head around it.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -0,0 +1,108 @@
+    * Note the above will only compile if there are [[Expansion]] instances for the types passed to the expand method.
+    */
+  protected class ExpandedMetricBuilder[A] private (m: String = "") {
+    def expand[A: Expansion](key: String, a: A) = {
```
The reuse of `A` here is giving a "suspicious shadowing by a Type Parameter" warning for me in IntelliJ.

rtitle replied:
Wow good catch! The type parameter on the class is totally unneeded and left over from an earlier incarnation of this.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -0,0 +1,108 @@
+
+  // Typeclass instances:
+
+  protected implicit object WorkspaceNameExpansion extends Expansion[WorkspaceName] {
```
comment here: overriding `makeName` to override default `toString` based implementation

rtitle replied:
👍

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -0,0 +1,108 @@
+    override def makeName(key: String, n: WorkspaceName): String = s"$key.${n.toString.replace('/', '.')}"
+  }
+
+  protected implicit object UUIDExpansion extends Expansion[UUID]
```
comment here: providing an implicit for UUIDs but can use the default `makeName`

rtitle replied:
👍

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -0,0 +1,108 @@
+
+  protected implicit object UUIDExpansion extends Expansion[UUID]
+
+  protected implicit def WorkflowStatusExpansion[A <: WorkflowStatus] = new Expansion[A] {}
```
why can't this be `implicit object WorkflowStatusExpansion extends Expansion[WorkflowStatus]`? that compiles for me

rtitle replied:
Ah, that works because we currently always pass the supertype `WorkflowStatus` to `ExpandedMetricBuilder`. E.g.:
```
// this compiles
private def workflowStatusCounter(status: WorkflowStatus): Counter =
  ExpandedMetricBuilder
    .expand(WorkspaceMetric, workspaceName)
    .expand(SubmissionMetric, submissionId)
    .expand(WorkflowStatusMetric, status)
    .asCounter()
```
But it doesn't work if you try to pass a subtype of `WorkflowStatus`:
```
// this does not compile: Cannot expand instances of type org.broadinstitute.dsde.rawls.model.WorkflowStatuses.Launching.type
private def foo: Counter = 
  ExpandedMetricBuilder
    .expand(WorkflowStatusMetric, WorkflowStatuses.Launching)
    .asCounter()
```
So that's why I added the upper type bound `A <: WorkflowStatus`. Originally I thought I could solve this by making the type parameter covariant (e.g. `sealed trait Expansion[+A]`) but that opened up a whole new can of worms.

So I think this is the cleaner solution, although maybe it could use a comment.

helgridly replied:
Great explanation, thanks!

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:
```diff
@@ -77,17 +106,50 @@ class SubmissionMonitorActor(val workspaceName: WorkspaceName,
+      WorkflowStatuses.allStatuses.foreach { status =>
+        ExpandedMetricBuilder
+          .expand(WorkspaceMetric, workspaceName)
+          .expand(SubmissionMetric, submissionId)
```
Can we pre-create and reuse as far as:

```
ExpandedMetricBuilder
    .expand(WorkspaceMetric, workspaceName)
    .expand(SubmissionMetric, submissionId)
```

since we're re-building this all over the place (not just here but every time `workflowStatusCounter` and `submissionStatusCounter` are called)?

Or am I prematurely optimising?

rtitle replied:
Yeah probably could reuse that. Though I don't think building the metric names is too expensive. :)

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:
```diff
@@ -77,17 +106,50 @@ class SubmissionMonitorActor(val workspaceName: WorkspaceName,
 
-  def scheduleNextMonitorPass: Cancellable = {
+  private def scheduleNextMonitorPass: Cancellable = {
     system.scheduler.scheduleOnce(submissionPollInterval, self, StartMonitorPass)
```
since you're nearby: this probably wants jitter too

rtitle replied:
👍

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:
```diff
@@ -107,36 +114,41 @@ trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths
   }
 
-  def reserveWorkflowBatch(dataAccess: DataAccess, activeCount: Int)(implicit executionContext: ExecutionContext): ReadWriteAction[Seq[WorkflowRecord]] = {
+  def reserveWorkflowBatch(dataAccess: DataAccess, activeCount: Int)(implicit executionContext: ExecutionContext): ReadWriteAction[Option[(Seq[WorkflowRecord], SubmissionRecord, WorkspaceRecord)]] = {
```
Blergh. I see why you need this - you need the submission and workspace details for `getUnlaunchedWorkflowBatch` - but this feels like a super heavyweight solution. I can't see another way around it though (sure, we could pass just the `submissionId` and `WorkspaceName` but it's still an option-of-triple).

rtitle replied:
+1 on Blergh. On the plus side it saves some database queries because we no longer need to re-lookup the `Workspace` and `Submission` in `submitWorkflowBatch` (since we remember them from workflow reservation time).

Maybe it could be made a bit cleaner with a container case class:
```
case class WorkflowBatch(workflowIds: Seq[Long], submissionRec: SubmissionRecord, workspaceRec: WorkspaceRecord)
```
and we pass that around instead of a `Tuple3`.

helgridly replied:
It's such a random glom of objects that making a case class just feels like denial. Might as well leave it as-is.

helgridly replied:
Oh, you already changed it. Never mind then.

jmthibault79 replied:
IMO random-glom-of-objects is a primary use case for case classes

rtitle commented:
Successfully tested pushing these metrics to stackdriver. See:

https://app.google.stackdriver.com/monitoring/1052499/rawls-metrics?project=broad-dsp-stackdriver
![screen shot 2017-07-05 at 6 01 14 pm](https://user-images.githubusercontent.com/5368863/27886891-ff124968-61ab-11e7-87ef-174101568d20.png)

They are appearing as expected as "flat" metrics in the SD UI. The next step will be to auto-create labels so we can do interesting aggregation in SD. (That will be done as part of the statsd/collectd plugin, not Rawls.)

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2146,6 +2157,19 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+  private def submissionStatusCounter(workspaceName: WorkspaceName, status: SubmissionStatus): Counter =
+    ExpandedMetricBuilder
+      .expand(WorkspaceMetric, workspaceName)
+      .expand(SubmissionStatusMetric, status)
```
submission id here?

rtitle replied:
Actually I was intending to count submission statuses per workspace, so it is not scoped to a single submission.

rtitle committed:
Code review comments, including moving wf/sub status counting to the *Component classes

rtitle committed:
Cleanup and comments

rtitle commented:
@helgridly I pushed code review feedback, including moving incrementing the counters to the `WorkflowComponent`/`SubmissionComponent` classes. Tests are still passing - I plan to repeat my FiaB test as well. Can you take another look? Thanks.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/metrics/RawlsInstrumented.scala:
```diff
@@ -95,10 +101,27 @@ trait RawlsInstrumented extends DefaultInstrumented {
+      .expand(SubmissionStatusMetric, status)
+      .asCounter()
+
+  protected def workflowStatusCounterProvider(workspaceName: WorkspaceName, submissionId: UUID) =
```
but this is `SubmissionMonitorActor.workflowStatusCounter` by another name, no? Can't we move that implicit somewhere shared and reuse it?

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:
```diff
@@ -202,7 +200,11 @@ trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths
-  def submitWorkflowBatch(workflowIds: Seq[Long], submissionRec: SubmissionRecord, workspaceRec: WorkspaceRecord)(implicit executionContext: ExecutionContext): Future[WorkflowSubmissionMessage] = {
+  def submitWorkflowBatch(batch: WorkflowBatch)(implicit executionContext: ExecutionContext): Future[WorkflowSubmissionMessage] = {
+
+    val WorkflowBatch(workflowIds, submissionRec, workspaceRec) = batch
```
TIL this offensively ugly syntax to do a useful thing.

helgridly commented:
Can we collapse `SubmissionMonitorActor.workflowStatusCounter` and `RawlsInstrumented.workflowStatusCounterProvider` down to the same object, presumably a shared implicit defined somewhere sensible? Same for `submissionStatusCounter` and any other similar friends.

👍 otherwise (and feel free to merge after the above change if my request is uncontroversial)

rtitle committed:
Code review feedback

General philosophy for counter definitions:
- RawlsInstrumented provides handy protected definitions for creating ExpandedMetricBuilders, counters, etc. These should be general and reusable, e.g. not tied to any business logic.
- Subclasses are repsonsible for instantiating the counters themselves
- If implicits are used to generate the counters, the implicits should live in the implementing classes close to the use site
